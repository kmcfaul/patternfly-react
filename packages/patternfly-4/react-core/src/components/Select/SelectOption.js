import React from 'react';
import styles from '@patternfly/patternfly/components/Select/select.css';
import { css } from '@patternfly/react-styles';
import PropTypes from 'prop-types';
import { SelectContext } from './selectConstants';

const propTypes = {
  /** the value for the option */
  children: PropTypes.string,
  /** additional classes added to the Select Option */
  className: PropTypes.string,
  /** the value for the option */
  value: PropTypes.string,
  /** flag indicating if the option is disabled */
  isDisabled: PropTypes.bool,
  /** flag indicating if the option acts as a placeholder */
  isPlaceholder: PropTypes.bool,
  /** Optional on click callback */
  onClick: PropTypes.func,
  /** Callback for ref tracking */
  sendRef: PropTypes.func,
  /** Additional props are spread to the container <button> */
  '': PropTypes.any
};

const defaultProps = {
  children: null,
  className: '',
  value: null,
  isDisabled: false,
  isPlaceholder: false,
  onClick: Function.prototype
};

class SelectOption extends React.Component {
  ref = React.createRef();

  componentDidMount() {
    this.props.sendRef(this.ref, this.props.index);
  }

  onKeyDown = event => {
    if (event.key === 'Tab') return;
    event.preventDefault();
    if (event.key === 'ArrowUp') {
      this.props.keyHandler(this.props.index, 'up');
    } else if (event.key === 'ArrowDown') {
      this.props.keyHandler(this.props.index, 'down');
    } else if (event.key === 'Enter') {
      this.ref.current.click && this.ref.current.click();
    }
  };

  render() {
    const {
      children,
      className,
      value,
      onClick,
      isDisabled,
      isPlaceholder,
      selected,
      sendRef,
      keyHandler,
      index,
      ...props
    } = this.props;
    return (
      <SelectContext.Consumer>
        {onSelect => (
          <li>
            <button
              {...props}
              className={css(
                styles.selectMenuItem,
                selected && styles.selectMenuItemMatch,
                isDisabled && styles.modifiers.disabled,
                className
              )}
              onClick={event => {
                if (!isDisabled) {
                  onClick && onClick(event);
                  onSelect && onSelect(event, value || children, isPlaceholder);
                }
              }}
              role="option"
              aria-selected={selected || null}
              ref={this.ref}
              onKeyDown={this.onKeyDown}
            >
              {value || children}
            </button>
          </li>
        )}
      </SelectContext.Consumer>
    );
  }
}

SelectOption.propTypes = propTypes;
SelectOption.defaultProps = defaultProps;

export default SelectOption;
