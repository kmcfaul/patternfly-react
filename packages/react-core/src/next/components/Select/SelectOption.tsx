import React from 'react';
import { css } from '@patternfly/react-styles';
import { MenuItemProps, MenuItem } from '../../../components/Menu';

export interface SelectOptionProps extends Omit<MenuItemProps, 'ref'> {
  /** Anything which can be rendered in a select option */
  children?: React.ReactNode;
  /** Classes applied to root element of select option */
  className?: string;
  /** Identifies the component in the Menu onSelect or onActionClick callback */
  itemId?: any;
  /** Flag indicating the option has a checkbox */
  hasCheck?: boolean;
  /** Render option as disabled */
  isDisabled?: boolean;
  /** Flag indicating if the option is selected */
  isSelected?: boolean;
  /** Flag indicating the option is focused */
  isFocused?: boolean;
}

export const SelectOption: React.FunctionComponent<MenuItemProps> = ({
  children,
  className,
  ...props
}: SelectOptionProps) => (
  <MenuItem className={css(className)} {...props}>
    {children}
  </MenuItem>
);
SelectOption.displayName = 'SelectOption';
