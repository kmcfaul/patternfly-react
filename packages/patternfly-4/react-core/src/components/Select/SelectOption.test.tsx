import React from 'react';
import { shallow } from 'enzyme';
import { SelectOption } from './SelectOption';

describe('select options', () => {
  test('renders with value parameter successfully', () => {
    const view = shallow(<SelectOption value="test" sendRef={jest.fn()} />);
    expect(view).toMatchSnapshot();
  });

  describe('disabled', () => {
    test('renders disabled successfully', () => {
      const view = shallow(<SelectOption isDisabled value="test" sendRef={jest.fn()} />);
      expect(view).toMatchSnapshot();
    });
  });

  describe('is selected', () => {
    test('renders selected successfully', () => {
      const view = shallow(<SelectOption isSelected value="test" sendRef={jest.fn()} />);
      expect(view).toMatchSnapshot();
    });
  });

  describe('checked', () => {
    test('renders with checked successfully', () => {
      const view = shallow(<SelectOption isChecked value="test" sendRef={jest.fn()} />);
      expect(view).toMatchSnapshot();
    });
  });
});
