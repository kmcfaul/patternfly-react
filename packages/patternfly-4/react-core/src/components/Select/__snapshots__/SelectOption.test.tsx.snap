// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select options checked renders with checked successfully 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`select options disabled renders disabled successfully 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`select options is selected renders selected successfully 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`select options renders with value parameter successfully 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
